Added L2 Regularization to Metalog OLS Fit Method

tjefferies opened the pull request:
Hey y'all! I added a new feature based on a conversation I had with Sam last week. See below for commit message.

Added an L2 Regularization term to the Metalog linear regression fit method, essentially turning the Metalog linear regression fit method into a Ridge Regression. This feature prevents overfitting and allows for the user to pick the highest term limit when using the returned metalog object with no degradation to performance.

Before Regularization:

![BeforeRegularization](https://user-images.githubusercontent.com/24706343/69008883-2355a500-0915-11ea-8d85-22a7d5220391.png)

After Regularization:

![AfterRegularization](https://user-images.githubusercontent.com/24706343/69008863-e7224480-0914-11ea-95d1-7efb2c5073ef.png)

If everything checks out I'll rebuild the package and republish it to pypi. Hope y'all are doing alright! Maybe we can jump on a call again soon and catch up?

tjefferies committed:
Added L2 Regularization to Metalog OLS Fit Method

Added an L2 Regularization term to the Metalog linear regression fit method, essentially turning the Metalog linear regression fit method into a Ridge Regression. This feature prevents overfitting and allows for the user to pick the highest term limit when using the returned metalog object with no degradation to performance.

colsmit commented:
Good work! Great idea, glad it got implemented. 

Minor quibble:  I suggest we rename the `regularization` parameter in `a_vector.a_vector_OLS_and_LP()` to something a little less ambiguous.
- One option would be to rename `regularization` to `l2-regularization`
- A more complete option would be to borrow from other packages and pass a `regularization = "l2"` or `penalty = "l2"` parameter as well as a `lambda = (x>=0)` parameter. Sklearn/liblinear pass the parameter `C` which is some ratio I don't understand. 

Would allow for adding other regularization later. I'm open to either or neither of these, I think its fine to merge as is. 

Would love to chat at some point, I started a new job recently and haven't had much of a chance to look at this, but things are starting to calm down a little.

tjefferies committed:
Added `penalty` Argument and Changed `regularization` Argument to `alpha`

Based on feedback from @colsmit, added a `penalty` argument and changed `regularization` term to `alpha` to align with sklearn.linear_model.Ridge class semantics. Going to create an issue and add a #TODO for L1 regularization fit method (LASSO) regression.

tjefferies commented:
@colsmit congrats on the new job! Good thinking on adding the `penalty` parameter. `lambda` is a keyword in base python so I ended up aligning the semantics with the [sklearn.linear_model.Ridge](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html) class and changed the `regularization` parameter to `alpha`.

`C` is used in `sklearn.linear_model.LogisticRegression` among other classes and is inverse lambda (inverse `alpha` in our case). I agree that the `C` semantic is confusing, hence why I decided to align it with the `sklearn.linear_model.Ridge` class.

Since we are adding a `penalty` parameter, I created an issue for both ElasticNet and L1 regularization (LASSO) regression. We should be ok using the `alpha` parameter for all three forms of regularization. L1 regularization is trickier than Ridge Regression as we have to formulate the fitting problem as an optimization problem. The [sklearn.linear_model.Lasso](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html) class implementation inherits from [sklearn.linear_model.ElasticNet](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNet.html) and uses coordinate descent to solve for the fit. sklearn.linear_model.ElasticNet uses an additional `l1_ratio` parameter to specify either L2 regularization (`l1_ratio=0`), L1 regularization (`l1_ratio=1`) or somewhere in between (`0 <= l1_ratio <= 1`). We can probably borrow their semantics for our implementation. If we can get the base ElasticNet feature working first, L1 regularization and L2 regularizations can be handled using a singular call to the ElasticNet functionality - probably worth benchmarking the coordinate descent approach vs the current numpy linear algebra operations first tho.

See 7bb171add0564637b4a634ab832485626ca5e092 for the updates.
